Update label definition regex

sukus21 opened the pull request:
This removes support for global labels without any colons at the end. I feel this is justified, as that style of label was removed from RGBDS in [v0.5.0](https://github.com/gbdev/rgbds/releases/tag/v0.5.0). That version released over 2 years ago, and the chances of anyone still using RGBDS v0.4.x at this point, are probably slim.

## Example
Here is an example of what label definitions are/aren't supported:

```asm
label          ;Not allowed, could overlap a macro invocation
label:         ;Allowed
label::        ;Allowed
.label         ;Allowed
.label:        ;Allowed
.label::       ;Allowed
label.local    ;Allowed
label.local:   ;Allowed
label.local::  ;Allowed
.label.local   ;Not allowed, sub-label within sub-label
.label.local:  ;Not allowed, sub-label within sub-label
.label.local:: ;Not allowed, sub-label within sub-label
```

## Benefits of this change
* This fixes macro documentation. Previously, when invoking a macro, the documentation provider would see the invokation as a label definition. This caused the documentation from the macro to be erased, and replaced with the documentation of the invocation of the macro. In the example below, the documentation provider would return `Invoke macro` for `name`, as it thinks it has been redeclared as a label:
```asm
; Loads a value into B.
MACRO name
    ld b, \1
ENDM

; Invoke macro
name 8
```

## Further considerations
I have considered doing a minor refactor of the documentation provider. The logic for macros and constants could be made simpler, as neither can be local or exported, and don't exactly belong to a scope either. That would free all of those terms for labels specifically, which opens up the ability to specialize them for that purpose.

DonaldHays commented on src/symbolDocumenter.ts:
```diff
@@ -16,7 +16,7 @@ const blockCommentEndRegex = /^(.*?)\s*\*\/.*$/
 const includeLineRegex = /^include[\s]+"([^"]+)".*$/i
 const spacerRegex = /^\s*(.)\1{3,}\s*$/
-const labelDefinitionRegex = /^[\s]*((([a-zA-Z_][a-zA-Z_0-9]*)?\.)?[a-zA-Z_][a-zA-Z_0-9]*[:]{0,2}).*$/
+const labelDefinitionRegex = /^\s*((?:[A-Z_]\w+)?(?:(?:\.[A-Z_]\w+:{0,2})|(?:[A-Z_]\w+:{1,2})))/i
```
```suggestion
const labelDefinitionRegex = /^\s*((?:[A-Z_]\w*)?(?:(?:\.[A-Z_]\w*:{0,2})|(?:[A-Z_]\w*:{1,2})))/i
```
In general this pull request looks like a great change, but I suggest changing the instances of `\w+` to `\w*`. Unless I'm mistaken, single-character label subcomponents are legal (e.g `m:` or `m.m::`), but I believe the regex you suggest requires at least two characters: one matching `[A-Z_]` and at least one additional character matching `\w+`, so it permits `mm:` but not `m:`.

In regards to breaking backwards compatibility: in general, I prefer to maintain support for deprecated and removed RGBDS features as long as possible, so long as doing so doesn't break anything. Since the original label regex interfered with macro use sites, this is a great example of the kind of situation that justifies breaking backwards compatibility. 👍

sukus21 replied:
That is a mistake from my side, thank you for catching that!

I agree with your stance on backwards compatibility, and I only removed support for the old label syntax, as I couldn't find a better, cleaner way to go about doing it. Although, as RGBDS is still under active development (pre-1.0), breaking changes from that side are somewhat inevitable. As far as I know, the community at large are pretty good at updating their RGBDS. Even the newer verisons of hardware.inc requires >0.5.0. These are the reasons that led me to justify removing support for the old label colon-less global label declaration.

sukus21 commented:
I rebased the branch on master, bumped the version number and updated the changelog.
Since this IS a breaking change, according to [semantic versioning](https://semver.org), should this require a new major version?

DonaldHays commented:
Yeah, I suppose a major version update will be justified.

I'll be planning to take a pass over the change log before actually making a release.
